feature: export image & multiple images

devangcx opened the pull request:
The PR includes the following features;

1. Change background color of a scene before exporting it to an image
2. Use ladybug tools like colors to shade geometries in the scene
3. Export an image with user specified size (width , height)
4. Extract radiance views from hbjson if found
5. Only export a certain elements from a model as an image
6. Export multiple images
7. Use parallel projection mode

resolves #116 
resolves #4

devangcx committed:
feat(scene): Add capability to assign camera and set output image size

devangcx committed:
test(scene_test): temporary test

devangcx committed:
chore(viewbased.hbjson): add a new sample file

devangcx committed:
chore(.gitignore): update gitignore

devangcx committed:
test(scene_test): update scene unit tests

devangcx committed:
test(scene_test): update scene tests

devangcx committed:
refactor(types): add a new method

devangcx committed:
feat(scene): bring Ladybug Tools colors to the image export

devangcx committed:
test(scene_test): update unit test

devangcx committed:
refactor(scene): add support for Camera object

devangcx committed:
feat(_helper): separate repeated functions in a helper module

devangcx committed:
feat(camera): add Camera object constructor

devangcx committed:
refactor(helper): remove hard-coded value length

devangcx committed:
feat(camera): add method to create cameras from radiance views

devangcx committed:
feat(model): extract radiance views from model

devangcx committed:
feat(scene): add multiple cameras to a scene

devangcx committed:
test(model_test): add unit test

devangcx committed:
test(scene_test): add unit tests

devangcx committed:
test(camera_test): add unit test for camera module

devangcx committed:
style(camera): update error message

devangcx committed:
chore(gridbased.hbjson): update hbjson with radiance views

devangcx committed:
feat(scene): add capability to export multiple images

devangcx committed:
test(scene_test): update unit tests and add tests

devangcx committed:
test(scene_test): update unit tests

devangcx committed:
refactor(scene): add error checking

devangcx committed:
feat(camera): add parallel projection and properties

devangcx committed:
refactor(camera): relax property checking

devangcx committed:
test(camera): add a unit test

devangcx committed:
fix(scene): add render method back

devangcx committed:
fix(scene): add render method back

devangcx committed:
style(scene_test): comment out two tests due to segmentation fault error on Github

devangcx committed:
chore(gridbased.hbjson): update sample file

devangcx committed:
fix(camera): fix parallel projection

devangcx committed:
refactor(scene): add method to find bounds

devangcx committed:
test(scene_test): update scene test suite

devangcx committed:
refactor(scene): move actors to separate module

devangcx committed:
feat(actors): add actors module

devngc commented:
I am in the middle of another short refactor. Hence, the PR is turned into a draft. It shoud not be reviewed while it's in the draft mode.

devangcx committed:
test(scene_test): update unit tests

devangcx committed:
test(actors_test): add unit tests for actors

mostaphaRoudsari reviewed (commented):
Hi @devngc, I know you are working on the PR but I made some comments yesterday. Here are my comments from yesterday for your information.

devangcx committed:
test(scene_test): update scene test

devangcx committed:
style(scene): improve error message

devangcx committed:
refactor(camera): add method to get adjusted camera position

devangcx committed:
style(actors): improve error messages

devangcx committed:
refactor(actors): turn staticmethod into an instance method

devangcx committed:
test(camera_test): add unit test for adjusted camera

devangcx committed:
feat(camera): add capability for adjusted camera position

devangcx committed:
test(scene_test): update unit tests

devangcx committed:
chore(gridbased.hbjson): update file

devangcx committed:
refactor(camera): improve method to get adjusted camera position

devangcx committed:
refactor(actors): make get_bounds return LB Point3Ds

devangcx committed:
refactor(camera): improve nearest point

devangcx committed:
refactor(actors): use to_vtk under the hood

devangcx committed:
refactor(camera): use to_vtk under the hood

devangcx committed:
test(scene): update scene tests

devangcx committed:
test(camera): update camera tests

devangcx committed:
refactor(scene): use to_vtk method under the hood

devangcx committed:
refactor(actors): add set_to_monochrome method

devangcx committed:
style(types): add missing spaces around operators

devangcx committed:
style(scene): update docstring and rename variables

devangcx committed:
refactor(model): load view from hbjson by default

devangcx committed:
refactor(camera): make Camera a sub-class of Radiance View

devangcx committed:
tests(actors): add unit tests for actors

devangcx committed:
tests(camera_test): add unit tests for camera

devangcx committed:
test(model_test): add unit tests for model

devangcx committed:
test(scene_test): update unit tests for scene

devngc commented on tests/scene_test.py:
```diff
@@ -91,6 +105,169 @@ def test_actors_in_scene():
+                    up_vector=(0.53, 0.40, 0.74), h_size=52.90)
+    scene = Scene(background_color=(255, 255, 255))
+
+    scene.add_cameras(camera.to_vtk())
```
Done.

devngc commented on honeybee_vtk/types.py:
```diff
@@ -374,6 +374,10 @@ def edge_visibility(self) -> bool:
 
+    def rgb_to_decimal(self):
+        """RGB color in decimal."""
+        return (self.color[0]/255, self.color[1]/255, self.color[2]/255)
```
Done.

devngc commented on honeybee_vtk/model.py:
```diff
@@ -63,8 +64,10 @@ def __init__(
                 honeybee-vtk model.
+            load_views: A boolean. If set to True, radiance views from the hbjson will
+                be loaded which can be used later to export images of the same views.
+                Defaults to False.
```
Done.

devngc commented on honeybee_vtk/model.py:
```diff
@@ -169,6 +182,14 @@ def _load_grids(self, model: HBModel, grid_options: SensorGridOptions):
+            return
+        elif hasattr(model.properties, 'radiance'):
+            for view in model.properties.radiance.views:
+                self._views.append(view)
```
Yes. I like that a Model object has views as its property. A Model object having cameras as its property would be confusing to me and probably others as well.

mostaphaRoudsari replied:
Hmm. I leave this to you but to me it's pretty normal for a VTK model to have cameras. Specially for a user who doesn't know the Honeybee model schema.

devngc replied:
Complied.

devngc commented on honeybee_vtk/_helper.py:
```diff
@@ -0,0 +1,28 @@
+        A boolean value if True or None.
+    """
+    # Check if all values in tuple are of expected object type
+    item_check = [isinstance(v, val_type) for v in val]
+
+    # Check if all values are less than the maximum allowed value
+    if max_val:
+        val_check = [v < max_val for v in val]
+    else:
+        val_check = [True] * len(val)
+
+    # final check
+    if item_check.count(True) == len(val) and val_check.count(True) == len(val):
+        return True
```
I did not want to create this method. I looked at the typing module in Honeybee-core first. See [this](https://github.com/ladybug-tools/honeybee-core/issues/418) issue.
1. I need a stronger type checking
2. I want to return a single boolean value rather than the items.

mostaphaRoudsari replied:
This is not a blocking change so if you like it to stay like this keep it but I can't get behind the logic here.

1. You can just use `map` functionality to do the same.
2. The code never really uses the boolean return. If you use maps it's more flexible and it give you similar results and you use the items that are created in the process.

devngc replied:
Implemented using the `map` function.

devngc commented on honeybee_vtk/camera.py:
```diff
@@ -0,0 +1,202 @@
+from honeybee.typing import clean_and_id_rad_string
+
+
+class Camera:
```
Done.

devngc commented on honeybee_vtk/camera.py:
```diff
@@ -0,0 +1,202 @@
+        if self._view_type == 'l':
+            fp = (self._position[0]+self._direction[0],
+                  self._position[1]+self._direction[1],
+                  self._position[2]+self._direction[2])
```
Done

devngc commented on honeybee_vtk/camera.py:
```diff
@@ -0,0 +1,202 @@
+            raise ValueError(
+                'Either load_views was not set to True while create a honeybee-vtk Model'
+                ' or no radiance views were found in the hbjson file.'
+            )
```
I don't understand. Why would we return an empty list? This from_model method is designed to work with model that has views. This error message should make it clear to the user that either drop using this method or bring a model with views.

mostaphaRoudsari replied:
Thinking about it more I think this classmethod should be `from_view` and not `from_model`. That will remove the need to check for such errors in the first place.

devngc replied:
Complied.

devngc commented on honeybee_vtk/camera.py:
```diff
@@ -0,0 +1,202 @@
+                    up_vector=view.up_vector.value,
+                    h_size=view.h_size.value,
+                    v_size=view.v_size.value,
+                    view_type=view.type.value).to_vtk() for view in model.views]
```
Done.

devngc commented on honeybee_vtk/scene.py:
```diff
@@ -20,93 +23,216 @@ class ImageTypes(enum.Enum):
+    Vtk derives inspiration from a movie set in naming its objects. Imagine a scene
+    being prepared at a movie set. The scene has a background. It has a few actors.
+    And, there are a few cameras setup to capture the scene from different angles. A
+    scene is vtk is defined in the similar fashion. It has a background color, some
```
Done.

devngc commented on honeybee_vtk/scene.py:
```diff
@@ -20,93 +23,216 @@ class ImageTypes(enum.Enum):
+    Args:
+        background_color: A tuple of three integers that represent RGB values of the
+            color that you'd like to set as the background color. Defaults to gray.
+        model: A Model camera object created using honeybee-vtk.
```
Done.

devngc commented on honeybee_vtk/scene.py:
```diff
@@ -20,93 +23,216 @@ class ImageTypes(enum.Enum):
+            the DisplayMode is set to Wireframe and results are going to be
+            loaded to the model.
+        monochrome_color: A tuple of decimal numbers to represent RGB color.
+            Defaults to gray.
```
Done.

devngc commented on honeybee_vtk/scene.py:
```diff
@@ -20,93 +23,216 @@ class ImageTypes(enum.Enum):
+        background_color: A tuple of three integers that represent RGB values of the
+            color that you'd like to set as the background color. Defaults to gray.
+        model: A Model camera object created using honeybee-vtk.
+        cameras: A list of vtk camera objects. Defaults to one camera that shows the
```
Imagine a scenario when a user wants to setup a camera using the constructor and want to add this camera to the scene in addition to the cameras she brings from the model. To provide this flexibility, I added cameras as a property to a scene. 

Also, in the real world of movies, if you want to shoot a scene, you setup cameras on the scene. So I feel tha this implementation is in alignment with a real world movie scene.

mostaphaRoudsari replied:
> Imagine a scenario when a user wants to setup a camera using the constructor and wants to add this camera to the scene in addition to the cameras she brings from a model. To provide this flexibility, I added cameras as a property to a scene.

I like that idea but it doesn't justify having it in the constructor.

```python
scene = Scene()
scene.add_camera(camera_1)
scene.add_camera(camera_2)
```

devngc replied:
Complied. Camera and Actors are removed from the constructor of the Scene object.

devngc commented on honeybee_vtk/scene.py:
```diff
@@ -20,93 +23,216 @@ class ImageTypes(enum.Enum):
+        if not val:
+            colors = vtk.vtkNamedColors()
+            self._background_color = colors.GetColor3d("SlateGray")
+        elif _check_tuple(val, int):
+            self._background_color = val
```
What about the following cases?
https://github.com/devngc/honeybee-vtk/blob/features/honeybee_vtk/scene.py#L95
https://github.com/devngc/honeybee-vtk/blob/features/honeybee_vtk/scene.py#L114
https://github.com/devngc/honeybee-vtk/blob/features/honeybee_vtk/camera.py#L63
https://github.com/devngc/honeybee-vtk/blob/features/honeybee_vtk/actors.py#L65

mostaphaRoudsari replied:
As I said above I don't see the value here but if you think it improves the code then let it be. I think a simple map function would do the same and is more flexible.

If we are keeping it you may want to rename the function. Your inputs are not necessary tuples. And check can mean so many things. The name of the function is really vague to me.

devngc replied:
Renamed the function.

devngc commented on honeybee_vtk/scene.py:
```diff
@@ -20,93 +23,216 @@ class ImageTypes(enum.Enum):
+        else:
+            raise ValueError(
+                f'A Model object created using honeybee-vtk required. Intead got {val}.'
+            )
```
Done.

devangcx committed:
style(camera): add comments

devangcx committed:
test(_helper_test): add unit tests for _helper module

devangcx committed:
test(scene_test): update unit tests

devangcx committed:
test(actors_test): update unit tests

devangcx committed:
test(_helper_test): add unit tests

devangcx committed:
refactor(_helper): Improve checking

devangcx committed:
style(_helper): make assertion more pythonic

mostaphaRoudsari commented on honeybee_vtk/actors.py:
```diff
@@ -0,0 +1,154 @@
+        model: A Model object created using honeybee-vtk Model.
+            Defaults to None.
+    """
+    def __init__(self, model):
```
After checking the rest of the code I think the input here shouldn't be a model. It should be a `ModelDataSet` - then you can have a `classmethod` to create this class `from_model`.

devngc replied:
Complied.

mostaphaRoudsari reviewed (commented):
Thank you @devngc! I like your approach for taking the model out of scene but I have a number of questions about some other aspects of the implementation.

devngc commented on honeybee_vtk/actors.py:
```diff
@@ -0,0 +1,154 @@
+    """
+    def __init__(self, model):
+        self.model = model
+        self._monochrome = False
+        self._monochrome_color = (0.54, 0.54, 0.54)
```
Complied.

devngc commented on honeybee_vtk/actors.py:
```diff
@@ -0,0 +1,154 @@
+        return actor
+
+    def get_bounds(self):
+        """Get A list of Ladybug Point3D objects that represent the bounds of actors.
```
The number of points depends on the geometry in the model so we don't know. Order to the points does not matter.

devngc commented on honeybee_vtk/camera.py:
```diff
@@ -0,0 +1,246 @@
+
+        self.bounds = bounds
+        self.camera_offset = camera_offset
+        self._flat_view_directions = {
+                (0, 0, -1): [2, '+'],
+                (0, 0, 1): [2, '-'],
+                (0, 1, 0): [1, '+'],
+                (0, -1, 0): [1, '-'],
+                (1, 0, 0): [0, '+'],
+                (-1, 0, 0): [0, '-'],
+            }
```
It is already documented in in `property` `flat_view_direction`

devngc commented on honeybee_vtk/camera.py:
```diff
@@ -0,0 +1,246 @@
+
+    @property
+    def bounds(self):
+        """Bounds of actors to be used in parallel projection."""
```
Complied. Removed bounds from the properties of a Camera object.

devngc commented on honeybee_vtk/camera.py:
```diff
@@ -0,0 +1,246 @@
+            )
+
+    @property
+    def camera_offset(self):
+        """Camera offset is a distance that will be kept between the outermost face of
+        the actors and adjusted camera position in parallel projection."""
```
Complied. Camera offset from the properties of a Camera object. Camera offset is the minimum distance that the adjusted camera will maintain from the geometry. For example, say you have a top camera that looks at the ground (XY plane). The highest point in your model is say (5,5,15). The camera offset in this case is the distance that adjusted camera will maintain from this top point. Currently, I am keeping the camera offset distance to 1 unit and I removed the option for a user to change this distance. I think we don't need to expose it but let me know if you think otherwise.

devngc commented on honeybee_vtk/camera.py:
```diff
@@ -0,0 +1,246 @@
+        if self._type.value == 'l':
+
+            # If a flat view in Parallel projection is requested
+            if self._direction.value in self._flat_view_directions:
+                if not self._bounds:
+                    raise ValueError(
+                        'Bounds of actors are required to generate one of the flat'
+                        ' views. Use get_bounds method of the Actors object to get'
+                        ' these bounds.'
+                    )
+                # get adjusted camera position
+                position = self._adjusted_position()
```
Complied.

devngc commented on honeybee_vtk/model.py:
```diff
@@ -54,7 +54,8 @@ class Model(object):
         self, model: HBModel,
-        load_grids: SensorGridOptions = SensorGridOptions.Ignore
+        load_grids: SensorGridOptions = SensorGridOptions.Ignore,
+        load_views: bool = True
```
I might be missing something here but I think I did as per our agreement. Please see that the load_view has default value of True.

devngc commented on honeybee_vtk/scene.py:
```diff
@@ -20,93 +20,172 @@ class ImageTypes(enum.Enum):
+    def actors(self, val):
+        if not val:
+            self._actors = None
+        elif isinstance(val, Actors):
```
This is a great idea. I implemented this.

devngc commented on honeybee_vtk/scene.py:
```diff
@@ -117,71 +196,34 @@ def get_legend(self, color_range) -> vtk.vtkScalarBarWidget():
 
         Args:
             folder: A valid path to where you'd like to write the gltf file.
+            renderer: A vtk renderer object created using create_render_window method.
```
Well, it can help if we create class for window. The same thing is happening at export_image method where we're using a window and an interactor object.

mostaphaRoudsari replied:
As long as there is a default window/renderer - I think the idea of having a class for window to make it easier for one to create a rendered makes a lot of sense.

devngc replied:
Revised. A new module is created.

devngc commented on honeybee_vtk/camera.py:
```diff
@@ -0,0 +1,246 @@
+
+
+    @classmethod
+    def from_model(cls, model):
```
@mostaphaRoudsari, do you mean that this method should be called in the Model's constructor and we should have a list of cameras as property of a Model object rather than views?

mostaphaRoudsari replied:
Yes. I think it's fine for a VTK Model to have `camera`s and not `view`s.

devngc replied:
Compplied.

devangcx committed:
refactor(actor): major refactor

1. The constructor accepts a ModelDataSet object now
2. Honeybee-vtk model can be used to derive actors using the from_model method

devangcx committed:
refactor(camera): remove bounds and camera_offset properties

devangcx committed:
refactor(scene): add methods to add and remove actors to the scene

devangcx committed:
test(scene_test): update unit tests

devangcx committed:
test(camera_test): update unit tests

devangcx committed:
test(actors_test): update actors unit tests

devangcx committed:
test(scene_test): remove two unit tests

devangcx committed:
refactor(*): refactor as per feedback

devangcx committed:
refactor(*): refactor as per feedback

devangcx committed:
refactor(*): refactor as per feedback

devangcx committed:
feat(window): add a new module

devangcx committed:
refactor(*): refactor as per feedback

devangcx committed:
refactor(*): refactor as per feedback

mostaphaRoudsari commented on honeybee_vtk/model.py:
```diff
@@ -54,7 +58,8 @@ class Model(object):
         self, model: HBModel,
-        load_grids: SensorGridOptions = SensorGridOptions.Ignore
+        load_grids: SensorGridOptions = SensorGridOptions.Ignore,
+        load_cameras: bool = True
```
I have a hard time to communicate this one! I think we should fully remove this option and ALWAYS load the cameras since they are small objects. That's why I think we probably don't need an option like this.

```suggestion
```

devngc replied:
This is clear now. Complied.

mostaphaRoudsari commented on honeybee_vtk/scene.py:
```diff
@@ -1,281 +1,177 @@
-        renderer.SetBackground(background_color)
-        renderer.TwoSidedLightingOn()
+    @property
+    def actors(self) -> Dict[str, Actor]:
+        """A dictionary of vtk actor name: vtk actor structure."""
+        return self._actors
```
Why do we need a dictionary? Can't we just have the name as a property of the Actor itself? That will be more elegant in my opinion.

devngc replied:
Complied.

mostaphaRoudsari commented on honeybee_vtk/scene.py:
```diff
@@ -1,281 +1,177 @@
-
-        return image_path.as_posix()
+            raise ValueError(
+                'Please add at least one camera to the scene.'
```
We don't usually use `please` in error messages. What about something like this: To be able to export an image a scene must at least have one camera assigned.

devngc replied:
Revised.

mostaphaRoudsari reviewed (approved):
Hi @devngc! I left 3 comments but overall I think it looks good to merge it in. Honestly, there are details that I'm not sure about but we can get to them later if they come up. The separation of `Window` class looks good but it will be ideal if you can find a more specific name like RendererWindow or something that doesn't get confused with a Window in a building.

devngc commented:
Thanks for all the patient reviews @mostaphaRoudsari. I renamed the module to "assistant". I think the name "assistant" goes well with the movie scene analogy used by VTK.

devangcx committed:
feat(assistant): add a new helper class

devangcx committed:
style(_helper): simplify a helper method

devangcx committed:
style(camera): fix a pep 8 issue

devangcx committed:
refactor(model): load cameras by default

devangcx committed:
refactor(scene): start using assistant object in scene

devangcx committed:
test(scene_test): update scene test

devangcx committed:
test(assistant_test): add a unit test for assistant

devangcx committed:
test(scene_test): update tests

devangcx committed:
test(model_test): update tests

devangcx committed:
test(camera_test): update tests

devangcx committed:
style(scene): Improve error reporting

devangcx committed:
style(assistant): improve error checking
